### PR TITLE
Implemented Classes Floor and Ceiling

### PR DESCRIPTION
**Pull request by ShikharJ**
As always, please review and suggest improvements.

### CODE REVIEW

File: symengine/functions.cpp [hunk 1]
```diff
@@ -329,6 +329,214 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+                or down_cast<const Infty &>(*arg).is_negative_infinity()) {
+                return arg;
+            }
+        }
+        if (is_a<Rational>(*arg)) {
+            if (down_cast<const Rational &>(*arg).is_int()) {
+                return arg;
+            }
+            throw NotImplementedError(
+                "Floor function for Rational class not yet implemented.");
```
Comment (ShikharJ) [inline symengine/functions.cpp]: @isuruf Is there a way of acquiring `num` and `den` as C++ `int`s ?

Comment (isuruf) [inline symengine/functions.cpp]: `Rational` supports arbitrary precision. They may not fit C++ `int`s.

Comment (ShikharJ) [inline symengine/functions.cpp]: In that case, how can this be expanded further?

**Comment by ShikharJ** [comment]
Ping @isuruf

File: symengine/functions.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/functions.cpp]: Use `mp_mod` functions for multiprecision integers.

File: symengine/functions.cpp [hunk 2]
```diff
@@ -329,6 +329,214 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+        }
+        if (is_a<RealDouble>(*arg)) {
```
Comment (isuruf) [inline symengine/functions.cpp]: This should be in `Evaluate`

File: symengine/complex_mpc.cpp [hunk 3]
```diff
@@ -908,6 +908,24 @@ class EvaluateMPC : public Evaluate
                 MPFR_RNDN);
         return complex_mpc(std::move(t));
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<ComplexMPC>(x))
+        mpc_class t(down_cast<const ComplexMPC &>(x).as_mpc().get_prec());
+        mpc_div_fr(t.get_mpc_t(),
+                   down_cast<const ComplexMPC &>(x).as_mpc().get_mpc_t(),
+                   MPFR_RNDDD);
```
Comment (ShikharJ) [inline symengine/complex_mpc.cpp]: @isuruf How can `MPFR_RNDDD` be made visible in `complex_mpc.cpp` ? I was unable to make out how the same was achieved for `MPFR_RNDN`.

Comment (isuruf) [inline symengine/complex_mpc.cpp]: Why do you need `MPFR_RNDDD` ? Isn't `MPFR_RNDN` enough?

Comment (ShikharJ) [inline symengine/complex_mpc.cpp]: As far as I'm able to understand from the docs, `mpc_div_fr` rounds up in the direction of `MPFR_RNDxy`, where `x` and `y` are the rounding criteria for real and imaginary parts. `MPFR_RNDN` would simply round to the nearest integer, but we are concerned with `floor` and `ceiling`.

Comment (isuruf) [inline symengine/complex_mpc.cpp]: Okay. It's `MPC_RNDDD` not `MPFR_RNDDD`

**Comment by ShikharJ** [comment]
This is ready for a review @isuruf.

File: symengine/real_mpfr.cpp [hunk 4]
```diff
@@ -1033,6 +1033,22 @@ class EvaluateMPFR : public Evaluate
         mpfr_exp(t.get_mpfr_t(), x_, MPFR_RNDN);
         return real_mpfr(std::move(t));
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_srcptr x_ = down_cast<const RealMPFR &>(x).i.get_mpfr_t();
+        mpfr_class t(mpfr_get_prec(x_));
+        mpfr_rint_floor(t.get_mpfr_t(), x_, MPFR_RNDD);
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: Shouldn't this be `mpfr_floor`?

File: symengine/real_mpfr.cpp [hunk 5]
```diff
@@ -1033,6 +1033,22 @@ class EvaluateMPFR : public Evaluate
...
+        return real_mpfr(std::move(t));
+    }
+    virtual RCP<const Basic> ceiling(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_srcptr x_ = down_cast<const RealMPFR &>(x).i.get_mpfr_t();
+        mpfr_class t(mpfr_get_prec(x_));
+        mpfr_rint_ceil(t.get_mpfr_t(), x_, MPFR_RNDU);
+        return real_mpfr(std::move(t));
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: Can you add a test for this?

File: symengine/complex_mpc.cpp [hunk 6]
```diff
@@ -908,6 +908,28 @@ class EvaluateMPC : public Evaluate
                 MPFR_RNDN);
         return complex_mpc(std::move(t));
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<ComplexMPC>(x))
+        mpc_class t(down_cast<const ComplexMPC &>(x).as_mpc().get_prec());
+        mpc_set_fr_fr(
```
Comment (isuruf) [inline symengine/complex_mpc.cpp]: This only sets the value and there's no rounding to an integer done here

File: symengine/functions.cpp [hunk 7]
```diff
@@ -329,6 +329,158 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+{
+    return floor(arg);
+}
+
+RCP<const Basic> floor(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (down_cast<const Number &>(*arg).is_exact()) {
+            if (is_a<Rational>(*arg)
+                and not down_cast<const Rational &>(*arg).is_int()) {
```
Comment (isuruf) [inline symengine/functions.cpp]: Where is this `is_int` coming from? Rationals are never integers in symengine.

File: symengine/functions.cpp [hunk 8]
```diff
@@ -329,6 +329,158 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+    return ceiling(arg);
+}
+
+RCP<const Basic> ceiling(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (down_cast<const Number &>(*arg).is_exact()) {
+            if (is_a<Rational>(*arg)
+                and not down_cast<const Rational &>(*arg).is_int()) {
+                return add(floor(arg), one);
```
Comment (isuruf) [inline symengine/functions.cpp]: Use `mp_cdiv_q`

File: symengine/real_double.cpp [hunk 9]
```diff
@@ -248,6 +248,16 @@ class EvaluateRealDouble : public EvaluateDouble<RealDouble>
             return number(std::log(std::complex<double>(d)));
         }
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return number(std::floor(down_cast<const RealDouble &>(x).i));
```
Comment (isuruf) [inline symengine/real_double.cpp]: SymPy returns an integer in this case

File: symengine/real_mpfr.cpp (hunk 4 above)
Comment (ShikharJ) [inline symengine/real_mpfr.cpp]: It should be, I wasn't sure of the double rounding requirement. It is clear to me now.

File: symengine/functions.cpp (hunk 7 above)
Comment (ShikharJ) [inline symengine/functions.cpp]: That is redundant. I'll remove it.

File: symengine/complex_mpc.cpp (hunk 6 above)
Comment (ShikharJ) [inline symengine/complex_mpc.cpp]: @isuruf What should be used here? I couldn't find `MPC`'s alternative to `floor` and `ceil`.

Comment (isuruf) [inline symengine/complex_mpc.cpp]: Use mpfr's on the real and imaginary parts

**Comment by isuruf** [comment]
SymPy returns an Integer for Floats. We should do the same.

File: symengine/functions.cpp [hunk 10]
```diff
@@ -329,6 +329,154 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+}
+
+RCP<const Basic> ceiling(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (down_cast<const Number &>(*arg).is_exact()) {
+            if (is_a<Rational>(*arg)) {
+                const Rational &s = down_cast<const Rational &>(*arg);
+                integer_class quotient;
+                mp_cdiv_q(quotient, SymEngine::get_num(s.as_rational_class()),
```
Comment (ShikharJ) [inline symengine/functions.cpp]: @isuruf `mp_cdiv_q` doesn't seem to be available under `Boost`. Can you suggest a possible workaround for this?

File: symengine/tests/basic/test_functions.cpp [hunk 11]
```diff
@@ -4081,3 +4085,147 @@ TEST_CASE("test_dummy", "[Dummy]")
...
+    CHECK(eq(*r, *integer(10)));
+    r = floor(real_mpfr(std::move(b)));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *integer(11)));
+#ifdef HAVE_SYMENGINE_MPC
+    mpc_class c(100);
+    mpc_set_fr_fr(c.get_mpc_t(), a.get_mpfr_t(), b.get_mpfr_t(), MPFR_RNDN);
+    r = floor(complex_mpc(std::move(c)));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *Complex::from_two_nums(*integer(10), *integer(11))));
```
Comment (ShikharJ) [inline symengine/tests/basic/test_functions.cpp]: @isuruf I couldn't make out why this case fails here, whereas, its counterpart in `ceiling` works flawlessly.

File: symengine/complex_mpc.cpp [hunk 12]
```diff
@@ -908,6 +908,38 @@ class EvaluateMPC : public Evaluate
...
+        mpfr_class t(down_cast<const ComplexMPC &>(x).as_mpc().get_prec());
+        mpfr_class u(down_cast<const ComplexMPC &>(x).as_mpc().get_prec());
+        mpfr_floor(
+            t.get_mpfr_t(),
+            mpc_realref(down_cast<const ComplexMPC &>(x).as_mpc().get_mpc_t()));
+        mpfr_floor(
+            u.get_mpfr_t(),
+            mpc_imagref(down_cast<const ComplexMPC &>(x).as_mpc().get_mpc_t()));
+        return Complex::from_two_nums(
+            *integer(numeric_cast<int>(mpfr_get_si(t.get_mpfr_t(), MPFR_RNDN))),
```
Comment (isuruf) [inline symengine/complex_mpc.cpp]: Why is this converting from a `mpfr_t` to an `signed int`?

File: symengine/functions.cpp [hunk 13]
```diff
@@ -329,6 +329,161 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+RCP<const Basic> floor(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (down_cast<const Number &>(*arg).is_exact()) {
+            if (is_a<Rational>(*arg)) {
+                const Rational &s = down_cast<const Rational &>(*arg);
+                integer_class quotient;
+                mp_fdiv_q(quotient, SymEngine::get_num(s.as_rational_class()),
+                          SymEngine::get_den(s.as_rational_class()));
+                return integer(quotient);
```
Comment (isuruf) [inline symengine/functions.cpp]: Use `integer(std::move(quotient))`

File: symengine/complex_mpc.cpp (hunk 12 above)
Comment (ShikharJ) [inline symengine/complex_mpc.cpp]: To do away with `floats`.

File: symengine/functions.cpp [hunk 14]
```diff
@@ -329,6 +329,161 @@ bool inverse_lookup(umap_basic_basic &d, const RCP<const Basic> &t,
...
+
+RCP<const Basic> ceiling(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg)) {
+        if (down_cast<const Number &>(*arg).is_exact()) {
+            if (is_a<Rational>(*arg)) {
+                const Rational &s = down_cast<const Rational &>(*arg);
+                integer_class quotient;
+#if SYMENGINE_INTEGER_CLASS != SYMENGINE_BOOSTMP
+                mp_cdiv_q(quotient, SymEngine::get_num(s.as_rational_class()),
```
Comment (isuruf) [inline symengine/functions.cpp]: This is not the right place for this. I don't mind implementing `cdiv` using `fdiv` in `mp_boost.cpp`.

File: symengine/real_double.cpp [hunk 15]
```diff
@@ -248,6 +248,18 @@ class EvaluateRealDouble : public EvaluateDouble<RealDouble>
             return number(std::log(std::complex<double>(d)));
         }
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        return integer(
+            numeric_cast<int>(std::floor(down_cast<const RealDouble &>(x).i)));
```
Comment (isuruf) [inline symengine/real_double.cpp]: `double` has 53 bits of precision, while `int` has only 32 bits.

File: symengine/real_mpfr.cpp [hunk 16]
```diff
@@ -1033,6 +1033,23 @@ class EvaluateMPFR : public Evaluate
         mpfr_exp(t.get_mpfr_t(), x_, MPFR_RNDN);
         return real_mpfr(std::move(t));
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_srcptr x_ = down_cast<const RealMPFR &>(x).i.get_mpfr_t();
+        mpfr_class t(mpfr_get_prec(x_));
+        mpfr_floor(t.get_mpfr_t(), x_);
+        return integer(mpfr_get_si(t.get_mpfr_t(), MPFR_RNDN));
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: Converting from mpfr to integer loses precision

File: symengine/complex_mpc.cpp (hunk 12 above)
Comment (isuruf) [inline symengine/complex_mpc.cpp]: You are losing precision here. `mpfr_t` has arbitrary precision and `signed int` has only 32 bits.

Comment (ShikharJ) [inline symengine/complex_mpc.cpp]: So then, should we let `mpfr` return `floats` ? Or would setting the default precision to 32 bits be a solution?

Comment (isuruf) [inline symengine/complex_mpc.cpp]: Convert `mpfr_t` to a `mpz_t`. http://www.mpfr.org/mpfr-current/mpfr.html#Conversion-Functions

File: symengine/tests/basic/test_functions.cpp (hunk 11 above)
Comment (ShikharJ) [inline symengine/tests/basic/test_functions.cpp]: Update: This still fails on https://travis-ci.org/ShikharJ/symengine/jobs/242915408 and https://travis-ci.org/ShikharJ/symengine/jobs/242915416 .

**Comment by ShikharJ** [comment]
```
test_floor
-------------------------------------------------------------------------------
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4089
...............................................................................
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4117: FAILED:
  CHECK( eq(*r, *integer(2)) )
with expansion:
  false
2 + 2*I
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4121: FAILED:
  CHECK( eq(*r, *Complex::from_two_nums(*integer(2), *integer(2))) )
with expansion:
  false
3
2
floor(x)
ceiling(x)
10
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4147: FAILED:
  CHECK( eq(*r, *integer(10)) )
with expansion:
  false
11
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4150: FAILED:
  CHECK( eq(*r, *integer(11)) )
with expansion:
  false
-------------------------------------------------------------------------------
test_ceiling
-------------------------------------------------------------------------------
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4161
...............................................................................
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4189: FAILED:
  CHECK( eq(*r, *integer(3)) )
with expansion:
  false
3 + 3*I
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4193: FAILED:
  CHECK( eq(*r, *Complex::from_two_nums(*integer(3), *integer(3))) )
with expansion:
  false
4
3
floor(x)
ceiling(x)
11
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4219: FAILED:
  CHECK( eq(*r, *integer(11)) )
with expansion:
  false
12
/home/travis/build/ShikharJ/symengine/symengine/tests/basic/test_functions.cpp:4222: FAILED:
  CHECK( eq(*r, *integer(12)) )
with expansion:
  false
===============================================================================
test cases:   60 |  58 passed | 2 failed
assertions: 1006 | 998 passed | 8 failed
```
All these fail on https://travis-ci.org/ShikharJ/symengine/jobs/242915409

File: symengine/real_double.cpp [hunk 17]
```diff
@@ -248,6 +248,20 @@ class EvaluateRealDouble : public EvaluateDouble<RealDouble>
             return number(std::log(std::complex<double>(d)));
         }
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealDouble>(x))
+        integer_class i;
+        mpz_set_d(get_mpz_t(i), std::floor(down_cast<const RealDouble &>(x).i));
```
Comment (ShikharJ) [inline symengine/real_double.cpp]: `mpz_set_d` and `get_mpz_t` are not available under `Boost`

Comment (isuruf) [inline symengine/real_double.cpp]: I think all of the integer classes can be constructed from a double.
Did you try,
```
integer_class i(std::floor(down_cast<const RealDouble &>(x).i));
```

Comment (ShikharJ) [inline symengine/real_double.cpp]: That returns
```
/home/shikhar/symengine/symengine/real_double.cpp:254:71: error: no matching function for call to ‘SymEngine::mpz_wrapper::mpz_wrapper(double)’
         integer_class i(std::floor(down_cast<const RealDouble &>(x).i));
```

Comment (isuruf) [inline symengine/real_double.cpp]: Okay. I think you should implement `mp_set_d` for all types.

File: symengine/mp_class.h [hunk 18]
```diff
@@ -711,6 +745,11 @@ inline double mp_get_d(const integer_class &i)
     return i.convert_to<double>();
 }
 
+inline void mp_set_d(integer_class &i, double a)
+{
+    boost::multiprecision::mpz_set_d(i, a);
```
Comment (ShikharJ) [inline symengine/mp_class.h]: @isuruf I couldn't find the `Boost`'s alternative to `mpz_set_d` either. How can I improve this?

Comment (isuruf) [inline symengine/mp_class.h]: `i = a` ?

Comment (ShikharJ) [inline symengine/mp_class.h]: That returns `/home/travis/build/ShikharJ/symengine/symengine/mp_class.h:750:9: error: no match for ‘operator=’ in ‘i = a’`

Comment (isuruf) [inline symengine/mp_class.h]: I've no idea. You'll have to look through the documentation and try locally

File: symengine/mp_class.h [hunk 19]
```diff
@@ -413,6 +431,11 @@ inline double mp_get_d(const piranha::integer &i)
     return mpz_get_d(i.get_mpz_view());
 }
 
+inline void mp_set_d(piranha::integer &i, double a)
+{
+    return mpz_set_d(i.get_mpz_view(), a);
```
Comment (isuruf) [inline symengine/mp_class.h]: There should be a way in piranha to do this. `i = a` or `i = piranha::integer(a)` might work.

File: symengine/mp_class.h (hunk 18 above)
Comment (isuruf) [inline symengine/mp_class.h]: http://www.boost.org/doc/libs/1_60_0/libs/multiprecision/doc/html/boost_multiprecision/tut/conversions.html

**Comment by ShikharJ** [comment]
I think this is ready for a review @isuruf.

File: symengine/complex_mpc.cpp [hunk 20]
```diff
@@ -908,6 +908,38 @@ class EvaluateMPC : public Evaluate
...
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<ComplexMPC>(x))
+        integer_class re, im;
+        mpfr_get_z(
+            get_mpz_t(re),
+            mpc_realref(down_cast<const ComplexMPC &>(x).as_mpc().get_mpc_t()),
+            MPFR_RNDD);
+        mpfr_get_z(
+            get_mpz_t(im),
```
Comment (isuruf) [inline symengine/complex_mpc.cpp]: If you use `get_mpz_t` for input, you need to do `mp_demote` afterwards. See `mp_class.h` for some examples.

File: symengine/real_mpfr.cpp [hunk 21]
```diff
@@ -1033,6 +1033,22 @@ class EvaluateMPFR : public Evaluate
...
+        integer_class i;
+        mpfr_get_z(get_mpz_t(i), x_, MPFR_RNDD);
+        return integer(std::move(i));
+    }
+    virtual RCP<const Basic> ceiling(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_srcptr x_ = down_cast<const RealMPFR &>(x).i.get_mpfr_t();
+        integer_class i;
+        mpfr_get_z(get_mpz_t(i), x_, MPFR_RNDU);
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: Same as before. If you use `get_mpz_t` for input, you need to demote.

File: symengine/real_mpfr.cpp [hunk 22]
```diff
@@ -1033,6 +1033,22 @@ class EvaluateMPFR : public Evaluate
         mpfr_exp(t.get_mpfr_t(), x_, MPFR_RNDN);
         return real_mpfr(std::move(t));
     }
+    virtual RCP<const Basic> floor(const Basic &x) const override
+    {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_srcptr x_ = down_cast<const RealMPFR &>(x).i.get_mpfr_t();
+        integer_class i;
+        mpfr_get_z(get_mpz_t(i), x_, MPFR_RNDD);
+        return integer(std::move(i));
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: Can you add a test for this?

**Comment by ShikharJ** [comment]
Ping @isuruf

File: symengine/printer.cpp [hunk 23]
```diff
@@ -887,6 +887,8 @@ std::vector<std::string> init_str_printer_names()
     names[DIRICHLET_ETA] = "dirichlet_eta";
     names[KRONECKERDELTA] = "kroneckerdelta";
     names[LEVICIVITA] = "levicivita";
+    names[FLOOR] = "floor";
+    names[CEILING] = "ceiling";
```
Comment (isuruf) [inline symengine/printer.cpp]: Can you add a test for this?

**Comment by isuruf** [comment]
@ShikharJ, can you resolve conflicts?

**Comment by ShikharJ** [comment]
Updated @isuruf.

**Comment by ShikharJ** [comment]
@isuruf Can this be merged? It is required for `Range` set.

**Comment by isuruf** [comment]
SymPy does one more additional thing for Add, `floor(1+x)` is simplified to `floor(x) + 1`. This can be done in a new PR.

**Pull request merged**